pyhf/pdf.py

```diff
@@ -131,6 +137,173 @@ def __init__(self, spec, **config_kwargs):
         # build up our representation of the specification
         self.config = _ModelConfig.from_spec(self.spec,**config_kwargs)
 
+        _allmods = []
+        _allsamples = []
+        _allchannels = []
+        _allmods = []
+        channel_nbins = {}
+
+        for c in self.spec['channels']:
+            _allchannels.append(c['name'])
+            for s in c['samples']:
+                channel_nbins[c['name']] = len(s['data'])
+                _allsamples.append(s['name'])
+                for mod in s['modifiers']:
+                    _allmods.append((mod['name'],mod['type']))
+        _allmods = list(set(_allmods))
+        _allsamples = list(set(_allsamples))
+        _allchannels = list(set(_allchannels))
+        self.do_samples  = _allsamples[:]
+        self.do_channels = _allchannels[:]
+        self.do_mods = _allmods[:]
+        self.channel_nbins = channel_nbins
+
+        self.finalized_stats = {k:finalize_stats(self.config.modifier(k)) for k,v in self.config.par_map.items() if 'staterror' in k}
+        self._make_mega()
+        self.prep()
+
+    def _make_mega(self):
+        helper = {}
+        for c in self.spec['channels']:
+            for s in c['samples']:
+                helper.setdefault(c['name'],{})[s['name']] = (c,s)
+
+        mega_mods = {}
+        import copy
+        for m,mtype in self.do_mods:
+            for s in self.do_samples:
+                modspec = {'type': mtype, 'name': m}
+                if mtype == 'histosys':
+                    modspec.setdefault('data',{})['hi_data'] = []
+                    modspec.setdefault('data',{})['lo_data'] = []
+                    modspec.setdefault('data',{})['mask'] = []
+                elif mtype == 'normsys':
+                    modspec.setdefault('data',{})['hi'] = []
+                    modspec.setdefault('data',{})['lo'] = []
+                    modspec.setdefault('data',{})['mask'] = []
+                elif mtype == 'normfactor':
+                    modspec.setdefault('data',{})['mask'] = []
+                elif mtype == 'shapefactor':
+                    modspec.setdefault('data',{})['mask'] = []
+                elif mtype == 'shapesys':
+                    modspec.setdefault('data',{})['mask'] = []
+                elif mtype == 'staterror':
+                    modspec.setdefault('data',{})['uncrt'] = []
+                    modspec.setdefault('data',{})['mask']  = []
+                mega_mods.setdefault(s,{})[m] = copy.deepcopy(modspec)
+                
```

Comment:
the mega-channel will consist of mega-samples that subscribe to mega-modifiers. i.e. while in normal histfactory, each sample might be affected by some modifiers and some not, here we change it so that samples are affected by *all modifiers*, but we set up the modifier data such that the application of the modifier does not actually change the bin value for bins that are not originally affected by that modifier